src/server/game/Handlers/ItemHandler.cpp

```diff
@@ -1250,3 +1229,153 @@ void WorldSession::HandleRemoveNewItem(WorldPackets::Item::RemoveNewItem& remove
+
+        ItemPosCountVec dest;
+        InventoryResult result = player->CanBankItem(reagent ? NULL_BAG : bag, NULL_SLOT, dest, item, false, true, reagent);
+
+        if (result == EQUIP_ERR_OK && !(dest.size() == 1 && dest[0].pos == src))
+        {
+            player->RemoveItem(item->GetBagSlot(), item->GetSlot(), true);
+            player->BankItem(dest, item, true);
+
+            return true;
+        }
+
+        return false;
+    }
+
+    // Loop through all the bags to see if it can be stack or not.
+    void StoreItemInBags(Player* player, Item* item)
+    {
+        if (StoreItemAndStack(player, item, INVENTORY_SLOT_BAG_0))
+            return;
+
+        for (uint32 i = INVENTORY_SLOT_ITEM_START; i < INVENTORY_SLOT_ITEM_END; i++)
+        {
+            if (StoreItemAndStack(player, item, i))
+                break;
+        }
+    }
+
+    void StoreItemInBank(Player* player, Item* item)
+    {
+        if (BankItemAndStack(player, item, NULL_SLOT))
+            return;
+
+        for (uint32 i = BANK_SLOT_BAG_START; i < BANK_SLOT_BAG_END; i++)
+        {
+            if (BankItemAndStack(player, item, i))
+                break;
+        }
+    }
+
+    void StoreItemInReagentBank(Player* player, Item* item)
+    {
+        if (BankItemAndStack(player, item, NULL_SLOT, true))
+            return;
+
+        for (uint32 i = REAGENT_SLOT_START; i < REAGENT_SLOT_END; i++)
+        {
+            if (BankItemAndStack(player, item, i, true))
+                break;
+        }
+    }
+
+    void SortBags(Player* player, void(Player::* fn)(std::function<bool(Player*, Item*, uint8 /*bag*/, uint8 /*slot*/)>&&))
+    {
+        // First pass to stack items in caller.
+        std::unordered_map<uint32, uint32> itemsQuality;
+        typedef std::multimap<uint32, Item*> SortItemsContainer;
+        SortItemsContainer items;
+
+        // Second pass, we collect the informations for sorting.
+        (player->*fn)([&items, &itemsQuality](Player* player, Item* item, uint8 /*bag*/, uint8 /*slot*/)
+        {
+            // We get the number of non-distinct items and item level for sorting.
+            items.insert(std::make_pair(item->GetEntry(), item));
+            itemsQuality[item->GetEntry()] = item->GetItemLevel(player);
+
+            return true;
+        });
+
+        // We get advantage of the multimap properties to sort our items.
+        std::multimap<uint32, SortItemsContainer::value_type> resultMap;
+        for (auto const& pair : items)
+            resultMap.insert(std::make_pair(itemsQuality[pair.first], pair));
+
+        // Third pass to swap all the items correctly.
+        auto itr = std::begin(resultMap);
+        (player->*fn)([&resultMap, &itr](Player* player, Item* /*item*/, uint8 bag, uint8 slot)
+        {
+            if (itr == std::end(resultMap))
+                return false;
+
+            uint16 pos = itr->second.second->GetPos();
+            player->SwapItem(pos, (bag << 8) | slot);
```

Comment:
Nit: wasn't there a macro or utility function to centralize this and avoid copy pastes in the unlikely event this math changes